### PR TITLE
Preserve the ESorts abstraction in the upper layers.

### PR DESCRIPTION
**Pull request by ppedrot**
We should never go back and forth between the low-level representation and the high level one that keeps a quotient relative to a universe state.

Overlays:
- https://github.com/coq-community/aac-tactics/pull/121
- https://github.com/LPCIC/coq-elpi/pull/409
- https://github.com/mattam82/Coq-Equations/pull/518
- https://gitlab.inria.fr/fbesson/itauto/-/merge_requests/7
- https://github.com/SkySkimmer/coq-lean-import/pull/2
- https://github.com/unicoq/unicoq/pull/76
- https://github.com/coq-community/paramcoq/pull/103
- https://github.com/MetaCoq/metacoq/pull/797

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

File: engine/eConstr.mli [hunk 1]
```diff
@@ -46,6 +46,14 @@ sig
   (** Returns the view into the current sort. Note that the kind of a variable
       may change if the unification state of the evar map changes. *)
 
+  val equal : Evd.evar_map -> t -> t -> bool
```
Comment (SkySkimmer) [inline engine/eConstr.mli:49]: Might be worth having a couple more apis like `ESorts.set` `ESorts.is_sprop`

Comment (ppedrot) [reply engine/eConstr.mli:49]: I thought about that but it's unclear if we can really leverage them in the long run. The parts that mess with universes currently do it in unsafe ways.

Comment (SkySkimmer) [reply engine/eConstr.mli:49]: I don't know what you mean by that. 

I just think `ESorts.is_prop sigma s` would be nicer to use than `Sorts.is_prop (ESorts.kind sigma s)`

Comment (ppedrot) [reply engine/eConstr.mli:49]: OK, OK, I will add some.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**